File: Python/Partie5/Exo22.py
import threading
import time
import random

compteur = 0
u = 1.0
mutex = threading.Lock()

def f1(x):
    return (x-1)**2 / 4.0

def f2(x):
    return (x-2)**2 / 6.0

def thread_T1():
    global compteur, u
    while True:
        with mutex:
            u = f1(u)
            compteur += 1
            print(f"T1: u = {u}, iterations = {compteur}")
        time.sleep(random.randint(1, 5))

def thread_T2():
    global compteur, u
    while True:
        with mutex:
            u = f2(u)
            compteur += 1
            print(f"T2: u = {u}, iterations = {compteur}")
        time.sleep(random.randint(1, 5))

t1 = threading.Thread(target=thread_T1, daemon=True)
t2 = threading.Thread(target=thread_T2, daemon=True)
t1.start()
t2.start()
try:
    while True:
        time.sleep(1)
except KeyboardInterrupt:
    pass

File: C/Partie5/Exo22.c
#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>
#include <unistd.h>
#include <time.h>

int compteur = 0;
double u = 1.0;
pthread_mutex_t mutex = PTHREAD_MUTEX_INITIALIZER;

double f1(double x)
{
    return (x - 1) * (x - 1) / 4.0;
}

double f2(double x)
{
    return (x - 2) * (x - 2) / 6.0;
}

void *thread_T1(void *arg)
{
    while (1)
    {
        pthread_mutex_lock(&mutex);
        u = f1(u);
        compteur++;
        printf("T1: u = %f, iterations = %d\n", u, compteur);
        pthread_mutex_unlock(&mutex);
        int duree = rand() % 5 + 1;
        sleep(duree);
    }
    return NULL;
}

void *thread_T2(void *arg)
{
    while (1)
    {
        pthread_mutex_lock(&mutex);
        u = f2(u);
        compteur++;
        printf("T2: u = %f, iterations = %d\n", u, compteur);
        pthread_mutex_unlock(&mutex);
        int duree = rand() % 5 + 1;
        sleep(duree);
    }
    return NULL;
}

int main()
{
    srand(time(NULL));
    pthread_t t1, t2;
    pthread_create(&t1, NULL, thread_T1, NULL);
    pthread_create(&t2, NULL, thread_T2, NULL);
    pthread_join(t1, NULL);
    pthread_join(t2, NULL);
    return 0;
}